File: app/src/main/java/com/pomo/sync/timer/ActivePhaseTimer.kt
package com.pomo.sync.timer

import java.util.UUID

/**
 * Kernel-backed live timer. Phase truth is the materializer projection; the
 * wall clock only derives remaining time for the accepted head.
 */
internal class ActivePhaseTimer(
    private val deviceId: String,
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) {
    private val facts = mutableListOf<TimerFact>()
    private var phaseAnchorMillis: Long = 0
    private var running: Boolean = false
    private var ownershipClaimId: String = ""

    val projection: ActivePhaseProjection
        get() = ActivePhaseMaterializer.materialize(facts)

    fun remainingMillis(): Long {
        val current = projection
        val plan = current.plan ?: return 0
        val head =
            facts
                .filter { it.operationId in current.heads }
                .maxByOrNull { it.operationId }
                ?: return plan.durationMillis
        val baseElapsed = head.elapsedMillis
        val live =
            if (running && head.action in setOf(TimerAction.START, TimerAction.RESUME)) {
                (nowMillis() - phaseAnchorMillis).coerceAtLeast(0)
            } else {
                0
            }
        return (plan.durationMillis - baseElapsed - live).coerceAtLeast(0)
    }

    fun start(
        plan: PhasePlan,
        phaseId: String = UUID.randomUUID().toString(),
    ): ActivePhaseProjection {
        facts.clear()
        running = true
        phaseAnchorMillis = nowMillis()
        ownershipClaimId = nextId()
        return append(
            TimerFact(
                operationId = nextId(),
                phaseId = phaseId,
                action = TimerAction.START,
                parentHeads = emptySet(),
                ownerDeviceId = deviceId,
                ownershipClaimId = ownershipClaimId,
                plan = plan,
                elapsedMillis = 0,
            ),
        )
    }

    fun pause(): ActivePhaseProjection {
        val current = requireActive()
        val elapsed = current.plan!!.durationMillis - remainingMillis()
        running = false
        return append(
            TimerFact(
                operationId = nextId(),
                phaseId = current.phaseId!!,
                action = TimerAction.PAUSE,
                parentHeads = current.heads,
                ownerDeviceId = deviceId,
                ownershipClaimId = ownershipClaimId,
                plan = current.plan,
                elapsedMillis = elapsed.coerceAtLeast(0),
            ),
        )
    }

    fun resume(): ActivePhaseProjection {
        val current = requireActive()
        running = true
        phaseAnchorMillis = nowMillis()
        return append(
            TimerFact(
                operationId = nextId(),
                phaseId = current.phaseId!!,
                action = TimerAction.RESUME,
                parentHeads = current.heads,
                ownerDeviceId = deviceId,
                ownershipClaimId = ownershipClaimId,
                plan = current.plan!!,
                elapsedMillis = current.plan.durationMillis - remainingMillis(),
            ),
        )
    }

    fun complete(): ActivePhaseProjection {
        val current = requireActive()
        running = false
        return append(
            TimerFact(
                operationId = nextId(),
                phaseId = current.phaseId!!,
                action = TimerAction.COMPLETE,
                parentHeads = current.heads,
                ownerDeviceId = deviceId,
                ownershipClaimId = ownershipClaimId,
                plan = current.plan!!,
                elapsedMillis = current.plan.durationMillis,
            ),
        )
    }

    private fun append(fact: TimerFact): ActivePhaseProjection {
        facts += fact
        return ActivePhaseMaterializer.materialize(facts)
    }

    private fun requireActive(): ActivePhaseProjection {
        val current = projection
        require(current.phaseId != null && current.plan != null) { "no active phase" }
        return current
    }

    private fun nextId(): String = UUID.randomUUID().toString().replace("-", "")
}

File: extension/src/sync/timer/activePhaseTimer.ts
import { materializeActivePhase, type ActivePhaseProjection, type PhasePlan, type TimerFact } from "./activePhase";

function nextId(): string {
  return crypto.randomUUID().replaceAll("-", "");
}

/**
 * Kernel-backed live timer. Phase truth is the materializer projection; the
 * wall clock only derives remaining time for the accepted head.
 */
export class ActivePhaseTimer {
  private readonly facts: TimerFact[] = [];
  private phaseAnchorMillis = 0;
  private running = false;
  private ownershipClaimId = "";

  constructor(
    private readonly deviceId: string,
    private readonly nowMillis: () => number = () => Date.now(),
  ) {}

  get projection(): ActivePhaseProjection {
    return materializeActivePhase(this.facts);
  }

  remainingMillis(): number {
    const current = this.projection;
    if (current.plan === null) return 0;
    const head = this.facts
      .filter((fact) => current.heads.has(fact.operationId))
      .sort((left, right) => left.operationId.localeCompare(right.operationId))
      .at(-1);
    if (head === undefined) return current.plan.durationMillis;
    const live = this.running && (head.action === "START" || head.action === "RESUME")
      ? Math.max(0, this.nowMillis() - this.phaseAnchorMillis)
      : 0;
    return Math.max(0, current.plan.durationMillis - head.elapsedMillis - live);
  }

  start(plan: PhasePlan, phaseId: string = crypto.randomUUID()): ActivePhaseProjection {
    this.facts.length = 0;
    this.running = true;
    this.phaseAnchorMillis = this.nowMillis();
    this.ownershipClaimId = nextId();
    return this.append({
      operationId: nextId(),
      phaseId,
      action: "START",
      parentHeads: new Set(),
      ownerDeviceId: this.deviceId,
      ownershipClaimId: this.ownershipClaimId,
      plan,
      elapsedMillis: 0,
      timeUncertain: false,
      composedElapsedMillis: null,
    });
  }

  pause(): ActivePhaseProjection {
    const current = this.requireActive();
    const elapsed = current.plan!.durationMillis - this.remainingMillis();
    this.running = false;
    return this.append({
      operationId: nextId(),
      phaseId: current.phaseId!,
      action: "PAUSE",
      parentHeads: current.heads,
      ownerDeviceId: this.deviceId,
      ownershipClaimId: this.ownershipClaimId,
      plan: current.plan!,
      elapsedMillis: Math.max(0, elapsed),
      timeUncertain: false,
      composedElapsedMillis: null,
    });
  }

  resume(): ActivePhaseProjection {
    const current = this.requireActive();
    this.running = true;
    this.phaseAnchorMillis = this.nowMillis();
    return this.append({
      operationId: nextId(),
      phaseId: current.phaseId!,
      action: "RESUME",
      parentHeads: current.heads,
      ownerDeviceId: this.deviceId,
      ownershipClaimId: this.ownershipClaimId,
      plan: current.plan!,
      elapsedMillis: current.plan!.durationMillis - this.remainingMillis(),
      timeUncertain: false,
      composedElapsedMillis: null,
    });
  }

  complete(): ActivePhaseProjection {
    const current = this.requireActive();
    this.running = false;
    return this.append({
      operationId: nextId(),
      phaseId: current.phaseId!,
      action: "COMPLETE",
      parentHeads: current.heads,
      ownerDeviceId: this.deviceId,
      ownershipClaimId: this.ownershipClaimId,
      plan: current.plan!,
      elapsedMillis: current.plan!.durationMillis,
      timeUncertain: false,
      composedElapsedMillis: null,
    });
  }

  private append(fact: TimerFact): ActivePhaseProjection {
    this.facts.push(fact);
    return materializeActivePhase(this.facts);
  }

  private requireActive(): ActivePhaseProjection {
    const current = this.projection;
    if (current.phaseId === null || current.plan === null) throw new Error("no active phase");
    return current;
  }
}
